IPython/core/magics/osm.py

```diff
@@ -160,65 +205,59 @@ def rehashx(self, parameter_s=''):
             os.environ.get('PATH','').split(os.pathsep)]
 
         syscmdlist = []
-        # Now define isexec in a cross platform manner.
-        if os.name == 'posix':
-            isexec = lambda fname:os.path.isfile(fname) and \
-                     os.access(fname,os.X_OK)
-        else:
-            try:
-                winext = os.environ['pathext'].replace(';','|').replace('.','')
-            except KeyError:
-                winext = 'exe|com|bat|py'
-            if 'py' not in winext:
-                winext += '|py'
-            execre = re.compile(r'(.*)\.(%s)$' % winext,re.IGNORECASE)
-            isexec = lambda fname:os.path.isfile(fname) and execre.match(fname)
         savedir = os.getcwd()
 
         # Now walk the paths looking for executables to alias.
         try:
             # write the whole loop for posix/Windows so we don't have an if in
             # the innermost part
-            if os.name == 'posix':
+            if self.is_posix:
+                print(path)
                 for pdir in path:
                     try:
                         os.chdir(pdir)
-                        dirlist = os.listdir(pdir)
                     except OSError:
                         continue
+
+                    # use with notation for python 3.6 onward
+                    dirlist = os.scandir(path=pdir)
                     for ff in dirlist:
-                        if isexec(ff):
+                        if self.isexec(ff):
+                            fname = ff.name
                             try:
                                 # Removes dots from the name since ipython
                                 # will assume names with dots to be python.
-                                if not self.shell.alias_manager.is_alias(ff):
+                                if not self.shell.alias_manager.is_alias(fname):
                                     self.shell.alias_manager.define_alias(
-                                        ff.replace('.',''), ff)
+                                        fname.replace('.',''), fname)
                             except InvalidAliasError:
                                 pass
                             else:
-                                syscmdlist.append(ff)
+                                syscmdlist.append(fname)
             else:
                 no_alias = Alias.blacklist
                 for pdir in path:
                     try:
                         os.chdir(pdir)
-                        dirlist = os.listdir(pdir)
                     except OSError:
                         continue
+
+                    # use with notation for python 3.6 onward
```

Comment:
I guess you mean types annotations that you removed , I thing that can be trimmed now.